serde/src/de/mod.rs

```diff
@@ -451,16 +451,36 @@ impl<'a> Display for Expected + 'a {
 /// [de]: https://docs.serde.rs/serde/de/index.html
 /// [codegen]: https://serde.rs/codegen.html
 /// [impl-deserialize]: https://serde.rs/impl-deserialize.html
-pub trait Deserialize: Sized {
+pub trait Deserialize<'de>: Sized {
     /// Deserialize this value from the given Serde deserializer.
     ///
     /// See the [Implementing `Deserialize`][impl-deserialize] section of the
     /// manual for more information about how to implement this method.
     ///
     /// [impl-deserialize]: https://serde.rs/impl-deserialize.html
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error> where D: Deserializer;
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where D: Deserializer<'de>;
 }
 
+/// A data structure that can be deserialized without borrowing any data from
+/// the deserializer.
+///
+/// This is primarily useful for trait bounds on functions. For example a
+/// `from_str` function may be able to deserialize a data structure that borrows
+/// from the input string, but a `from_reader` function may only deserialize
+/// owned data.
+///
+/// ```rust,ignore
+/// pub fn from_str<'a, T>(s: &'a str) -> Result<T>
+///     where T: Deserialize<'a>;
+///
+/// pub fn from_reader<R, T>(rdr: R) -> Result<T>
+///     where R: Read,
+///           T: DeserializeOwned;
+/// ```
+pub trait DeserializeOwned: for<'de> Deserialize<'de> {}
```

Comment:
I understand that difference, but I mean in practical terms they aren't different, because in both cases, whatever type you deserialize to can't have any lifetimes other than `'static` (or must be something like `Option::<&'a Foo>::None`, which again, would turn `'a` into `'static`, since you can turn it into any other lifetime you need afterwards).

To remove all confusion: Can you think of any type that can be deserialized in the `for<'de> Deserialize<'de>` scheme, but not in the `Deserialize<'static>` scheme?